modelforge/potential/sake.py

```diff
@@ -0,0 +1,382 @@
+            Stability constant to prevent numerical instabilities (default is 1e-8).
+
+        Returns
+        -------
+        Tuple[torch.Tensor, torch.Tensor]
+            Updated scalar and vector representations (q, mu).
+        """
+        idx_i, idx_j = pairlist
+        nr_of_atoms_in_all_systems, _ = x.shape
+        r_ij = x[idx_j] - x[idx_i]
+        d_ij = torch.norm(r_ij, dim=-1)
+        dir_ij = r_ij / (d_ij.unsqueeze(-1) + self.epsilon)
+
+        h_ij_edge = self.edge_model(h[idx_i], h[idx_j], d_ij)
+        h_ij_semantic = self.get_semantic_attention(h_ij_edge, idx_i, d_ij, nr_of_atoms_in_all_systems)
+        h_i_semantic = self.aggregate(h_ij_semantic, idx_i, nr_of_atoms_in_all_systems)
+        combinations = self.get_combinations(h_ij_semantic, dir_ij)
+        h_i_spatial = self.get_spatial_attention(combinations, idx_i, dir_ij, nr_of_atoms_in_all_systems)
+        h_updated = self.update_node(h, h_i_semantic, h_i_spatial)
+        v_updated = self.update_velocity(v, h, combinations, idx_i)
+        x_updated = x + v_updated
+
+        return h_updated, x_updated, v_updated
+
+
+class LightningSAKE(SAKE, LightningModuleMixin):
+    def __init__(
+            self,
+            embedding: nn.Module,
+            nr_interaction_blocks: int,
+            nr_heads: int,
+            radial_basis: nn.Module,
+            cutoff: nn.Module,
+            activation: Optional[Callable] = F.silu,
+            loss: Type[nn.Module] = nn.MSELoss(),
+            optimizer: Type[torch.optim.Optimizer] = torch.optim.Adam,
+            lr: float = 1e-3,
+    ) -> None:
+        """PyTorch Lightning version of the SAKE model."""
+
+        super().__init__(
+            embedding_module=embedding,
+            nr_interaction_blocks=nr_interaction_blocks,
+            nr_heads=nr_heads,
+            radial_basis_module=radial_basis,
+            cutoff_module=cutoff,
+            activation=activation,
+        )
+        self.loss_function = loss
+        self.optimizer = optimizer
+        self.learning_rate = lr
+
+
+class ContinuousFilterConvolutionWithConcatenation(nn.Module):
+
+    def __init__(self, in_features, hidden_features, out_features, radial_basis_module, activation):
+        super().__init__()
+        self.rbf = radial_basis_module
+        self.mlp_in = nn.Linear(in_features, radial_basis_module.n_rbf)
+        self.mlp_out = nn.Sequential(
+            Dense(in_features + radial_basis_module.n_rbf + 1, hidden_features, activation=activation),
+            nn.Linear(hidden_features, out_features),
+        )
+
+    def forward(self, h_i, h_j, d_ij):
+        h_ij_cat = torch.cat([h_i, h_j], dim=-1)
+        h_ij_filtered = self.rbf(d_ij) * self.mlp_in(h_ij_cat)
+
+        h_ij_out = self.mlp_out(
+            torch.cat(
+                [h_ij_cat, h_ij_filtered, d_ij.unsqueeze(-1)],
+                dim=1
+            )
+        )
+
+        return h_ij_out
+
+
+class ExpNormalSmearing(nn.Module):
```

Comment:
Can you please add docstrings explaining the functionality and use case, and the parameters that are passed?